### PR TITLE
Linear Regression using Analytical approach

### PR DESCRIPTION
**Pull request by somnat**

### CODE REVIEW

**Comment by markroxor** [comment]
- [ ] write docstrings.
- [ ] fix pep8 errors.
- [ ] use pytorch.
- [ ] use logger, no print statements
- [ ] write tests.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+            for i in range(len(test_data)):
+                predicted_value = self.b0 + self.b1 * test_data[i]
+                logger.info("Predicted value for test_data {} slope {} and bias {} is {}".format(test_data[i], self.b1,
+                            self.b0, predicted_value))
+        if self.option == 'SGD':
+            for i in range(len(test_data)):
+                predicted_value = self.alpha + self.beta * test_data[i]
+                logger.info("Predicted value for test_data {} slope {} and bias {} is {}".format(test_data[i],
+                            self.beta, self.alpha, predicted_value))
+        if self.option == 'Adam':
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: Why do we need this `if` if both the options do the exact same thing.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+                self.cov_x_y.append((self.x[i] - mean_x) * (self.y[i] - mean_y))
+                self.var_x.append((self.x[i] - mean_x) ** 2)
+            self.b1 = sum(self.cov_x_y) / sum(self.var_x)
+            self.b0 = mean_y - self.b1 * mean_x
+        if option == 'SGD':
+            x_data = Variable(ch.Tensor(self.x), requires_grad=False)
+            y_data = Variable(ch.Tensor(self.y), requires_grad=False)
+            self.beta = Variable(ch.randn(1, 1), requires_grad=True)
+            self.alpha = Variable(ch.randn(1), requires_grad=True)
+            optimizer = ch.optim.SGD([self.beta, self.alpha], lr=0.01)
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: We need to write our own optimizers and limit the use of `pytorch` for "numerical operations" only. We need to define a helper function for these optimizers.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+import torch as ch
+import logging
+from torch.autograd import Variable
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: We cannot use torch's `autograd` as this is not a `numpy` feature.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+                loss = (y_pred - y_data).pow(2).sum()
+                if i % 50 == 0:
+                    logger.info("iteration {} loss {}".format(i, loss.data[0]))
+                    optimizer.zero_grad()
+                    loss.backward()
+                    optimizer.step()
+
+    def predict(self, test_data):
+        if self.option == 'Analytical':
+            for i in range(len(test_data)):
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: Consider using -
``` python
for td in test_data:
```

File: omega/linear_regression/test_linear_regression.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        # compares two omg.nn models by comparing their weights and biases
+        for wt1, wt2 in zip(m1.layerwise_weights, m2.layerwise_weights):
+            self.assertTrue(torch.equal(wt1, wt2))
+
+        for b1, b2 in zip(m1.layerwise_biases, m2.layerwise_biases):
+            self.assertTrue(torch.equal(b1, b2))
+
+    def test_consistency(self):
+        # tests for model's load save consistency.
+        old_nw = omg.nn.NetworkMesh([2, 5, 2], seed=100)
```
Comment (somnat) [inline omega/linear_regression/test_linear_regression.py]: Modify the test_consistency module. I think it is not generic and suitable for all sort of learning technique. I

Comment (markroxor) [reply omega/linear_regression/test_linear_regression.py]: It is not supposed to be generic. It is specific to each model.

File: omega/linear_regression/LinearRegressionClassifier.py (hunk 3 above)
Comment (markroxor) [reply omega/linear_regression/LinearRegressionClassifier.py]: Please change this.

**Comment by markroxor** [comment]
I don't find the `linear_regression.rst` file which I created for you. It is important as it will show the docstrings on the official website. Please commit it.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+            self.var_x = []
+            mean_x = ch.mean(self.x)
+            mean_y = ch.mean(self.y)
+            for i in range(len(self.x)):
+                self.cov_x_y.append((self.x[i] - mean_x) * (self.y[i] - mean_y))
+                self.var_x.append((self.x[i] - mean_x) ** 2)
+            self.beta = sum(self.cov_x_y) / sum(self.var_x)
+            self.alpha = mean_y - self.beta * mean_x
+            return
+        ch.manual_seed(2)
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: Please dont hard code the `seed`. Check out `nn`'s implementation for the use of seed.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+        x_data = Variable(ch.FloatTensor(self.x), requires_grad=False)
+        y_data = Variable(ch.FloatTensor(self.y), requires_grad=False)
+        self.beta = Variable(ch.randn(1, 1), requires_grad=True)
+        self.alpha = Variable(ch.randn(1), requires_grad=True)
+
+        if option == 'SGD':
+            optimizer = ch.optim.SGD([self.beta, self.alpha], lr=0.01)
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: We need to write our own optimizers. Please remove this.

File: omega/linear_regression/test_linear_regression.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
```
Comment (markroxor) [inline omega/linear_regression/test_linear_regression.py]: This file belongs to `omega/tests` being outside the test folder the tests in this file will not be executed.

File: omega/linear_regression/test_linear_regression.py [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+from omega.test.toolbox import _tempfile, _test_data_path
+
+import logging
+
+logging.basicConfig()
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class TestNN(unittest.TestCase):
```
Comment (markroxor) [inline omega/linear_regression/test_linear_regression.py]: Please dont push files which are not intended to be merged.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+        elif option == 'Adam':
+            optimizer = ch.optim.Adam([self.beta, self.alpha], lr=0.01)
+
+        for i in range(1000):
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: Please avoid hard coding the values.

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+   Examples
+   --------
+   >>> import omega as omg
+   >>> import torch as ch
+   >>> x = ch.Tensor([[24], [50], [15], [55], [14], [12], [18], [62]])
+   >>> y = ch.Tensor([[21.54945196], [47.46446305], [17.21865634], [52.789],
+                    [16.1234], [12.789], [19.5649], [60.5793278]])
+   >>> lr = omg.linear_regression.LinearRegressionClassifier(x,y)
+   >>> lr.fit('Adam')
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: change this to "Analytical"

File: omega/linear_regression/LinearRegressionClassifier.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+            Parameters
+            ----------
+            test_data : list of (int/float torch.Tensor).
+                The test data on which the results are evaluated generally after each
+                epoch.
+
+            """
+            if not hasattr(self, "alpha") or not hasattr(self, "beta"):
+                raise NotImplementedError("Predict is called before fit."
+                                          "Did you wish to load the load the model?")
```
Comment (markroxor) [inline omega/linear_regression/LinearRegressionClassifier.py]: Fix this statement.

File: omega/test/test_linear_regression.py [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+
+class TestLinearRegression(unittest.TestCase):
+    def setUp(self):
+        # sets up a basic input dataset which implements a XOR gate.
+        self.x = ch.Tensor([[1.0], [2.0], [3.0]])
+        self.y = ch.Tensor([[2.0], [4.0], [6.0]])
+
+    def model_equal(self, m1, m2):
+        # compares two omg.nn models by comparing their weights and biases
+        print(type(m1.alpha), m1.alpha)
```
Comment (markroxor) [inline omega/test/test_linear_regression.py]: remove this `print`

**Comment by markroxor** [comment]
Dr. @somnat many thanks for your efforts.
